src/dist/replication/lib/replica_2pc.cpp

```diff
@@ -130,6 +150,22 @@ void replica::init_prepare(mutation_ptr &mu, bool reconciliation)
         goto ErrOut;
     }
 
+    // stop prepare bulk load ingestion if there are secondaries unalive
+    for (auto i = 0; i < request_count; ++i) {
+        const mutation_update &update = mu->data.updates[i];
+        if (update.code == dsn::apps::RPC_RRDB_RRDB_BULK_LOAD) {
+            ddebug_replica("try to prepare bulk load mutation({})", mu->name());
+            if (static_cast<int>(_primary_states.membership.secondaries.size()) + 1 <
+                _primary_states.membership.max_replica_count) {
+                err = ERR_NOT_ENOUGH_MEMBER;
+                break;
+            }
+        }
```

Comment:
If one update in updates is not RPC_RRDB_RRDB_BULK_LOAD type, the others should not be RPC_RRDB_RRDB_BULK_LOAD type too? If that, you can break the loop early.

Reply:
For RPC_RRDB_RRDB_BULK_LOAD, it can not be batched, so this mutation will only have one update.

Reply:
I mean if found a non-RPC_RRDB_RRDB_BULK_LOAD 'update' in the loop, we can break the loop to avoid meanless loop (because non-RPC_RRDB_RRDB_BULK_LOAD mutation is the most normal case and it may have multiple 'updates').
 > so this mutation will only have one update.

You can add an `assert` to ensure it.

Reply:
oh, I got your point. I have updated the code. 
If a write operation is defined as batch_not_allowed, such as multi_set, the mutation will only have one client request, and one update in mutation's data field, it is guaranteed by other function. I think it's okay for not adding an assert here.